### PR TITLE
Update Sentinel-5P

### PR DESCRIPTION
**Pull request by pjhartzell**
## Description

Adds the `sentinel-5p-l2-netcdf` collection to the `sentinel-5p` dataset. 

The Sentinel-5P payload is a single imaging spectrometer, the TROPOMI sensor. The collected data are processed to Level-2 derived products, 13 of which are included in this PR's `sentinel-5p-l2-netcdf` collection. The products are listed below, _each linked to a sample STAC Item_ in PC Test:

* [`L2__AER_AI`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_AER_AI_20180628T102407_20180628T120537_03661): Ultraviolet aerosol index
* [`L2__AER_LH`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_AER_LH_20180430T185522_20180430T203851_02829): Aerosol layer height
* [`L2__CH4___`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_CH4____20180430T001851_20180430T020219_02818): Methane (CH<sub>4</sub>) total column
* [`L2__CLOUD_`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_CLOUD__20180430T020021_20180430T034349_02819): Cloud fraction, albedo, and top pressure
* [`L2__CO____`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_CO_____20180430T001851_20180430T020219_02818): Carbon monoxide (CO) total column
* [`L2__HCHO__`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_HCHO___20180514T211549_20180514T225917_03029): Formaldehyde (HCHO) total column
* [`L2__NO2___`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_NO2____20180430T001851_20180430T020219_02818): Nitrogen dioxide (NO<sub>2</sub>) total column
* [`L2__O3____`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_O3_____20180430T001851_20180430T020219_02818): Ozone (O<sub>3</sub>) total column
* [`L2__O3_TCL`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_O3_TCL_20180430T105026_20180506T131804_02824): Ozone (O<sub>3</sub>) tropospheric column
* [`L2__SO2___`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_SO2____20180506T170136_20180506T184419_02913): Sulfur dioxide (SO<sub>2</sub>) total column
* [`L2__NP_BD3`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_NP_BD3_20180430T001851_20180430T020219_02818): Cloud from the Suomi NPP mission, band 3
* [`L2__NP_BD6`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_NP_BD6_20180430T001851_20180430T020219_02818): Cloud from the Suomi NPP mission, band 6
* [`L2__NP_BD7`](https://pct-apis-staging.westeurope.cloudapp.azure.com/stac/collections/sentinel-5p-l2-netcdf/items/S5P_L2_NP_BD7_20180430T001851_20180430T020219_02818): Cloud from the Suomi NPP mission, band 7

You can check out the [dataset landing page in PC Test](https://planetarycomputer-test.microsoft.com/dataset/sentinel-5p-l2-netcdf).

## Notes

* The STAC Items are derived from the existing STAC Items on Azure. The NetCDF files are not touched.
* The products are still in swath form (not re-gridded to tiles)
* The products are in NetCDF format
* There are 14 products in Azure, but we are only indexing 13 of them with STAC. The product that does not have STAC Items in Azure is `L2__O3__PR`, which is an ozone profile product. The data in Azure for this product starts in November 2021, so perhaps this product became available after the STAC Item creation pipeline was put together.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

The Collection and several hundred Items were ingested into PC Test.

## Checklist:

Please delete options that are not relevant.

- [ ] I have performed a self-review
- [ ] Changelog has been updated
- [ ] Documentation has been updated
- [ ] Unit tests pass locally (./scripts/test)
- [x] Code is linted and styled (./scripts/format)

### CODE REVIEW

File: datasets/sentinel-5p/collection/template.json [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+            "rel": "about",
+            "href": "https://sentinel.esa.int/web/sentinel/missions/sentinel-5p",
+            "type": "text/html",
+            "title": "Sentinel-5 Precursor Mission"
+        }
+    ],
+    "stac_extensions": [
+        "https://stac-extensions.github.io/sat/v1.0.0/schema.json",
+        "https://stac-extensions.github.io/table/v1.2.0/schema.json"
+    ],
```
Comment (gadomski) [inline datasets/sentinel-5p/collection/template.json:25]: Needs `item-assets` extension.

File: datasets/sentinel-5p/collection/template.json [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        }
+    ],
+    "assets": {
+        "thumbnail": {
+            "title": "Sentinel-5P Level-2 NetCDF",
+            "href": "https://ai4edatasetspublicassets.blob.core.windows.net/assets/pc_thumbnails/sentinel-5p-l2-netcdf-thumb.png",
+            "media_type": "image/png"
+        },
+        "geoparquet-items": {
+            "href": "abfs://items/io-biodiversity.parquet",
```
Comment (gadomski) [inline datasets/sentinel-5p/collection/template.json:63]:
```suggestion
            "href": "abfs://items/sentinel-5p-l2a-netcdf.parquet",
```

**Review by gadomski — CHANGES_REQUESTED** [review]

File: datasets/sentinel-5p/collection/template.json (hunk 1 above)
Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json]: Good catch. Thanks.

File: datasets/sentinel-5p/collection/template.json (hunk 2 above)
Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json]: Ugh. The dangers of copy/paste. Thanks for catching that.

**Review by gadomski — APPROVED** [review]

**Comment by TomAugspurger** [comment]
@pjhartzell sorry, I caused a merge conflict with #166. That just added the `msft:region` property to the file collection JSON you moved / updated.

Can you comment on what the `summaries` JSON files are for? e.g. datasets/sentinel-5p/summaries/2022-09-08-full.json. They probably shouldn't be committed, right?

**Comment by pjhartzell** [comment]
@TomAugspurger 
>sorry, I caused a merge conflict with https://github.com/microsoft/planetary-computer-tasks/pull/166. That just added the msft:region property to the file collection JSON you moved / updated.

No worries. 

>Can you comment on what the summaries JSON files are for? e.g. datasets/sentinel-5p/summaries/2022-09-08-full.json. They probably shouldn't be committed, right?

I suppose not. These files existed in the sentinel-5p folder before I started this PR. I'll move them out.

**Comment by TomAugspurger** [comment]
Ah, thanks, I missed that they were a move. I suspect that Rob made those when analyzing the existing STAC items to find good values for summaries.

**Comment by pjhartzell** [comment]
Yep, and they were valuable in both content and as an example of generating summary info. I'll hold onto them locally.

**Review by TomAugspurger — COMMENTED** [review]
Just confirm: we need to add a bit of code here to call the new `enclose_poles`?

FWIW, I tried it out on the asset at `blob://sentinel5euwest/sentinel-5p-stac/TROPOMI/L2__AER_AI/2023/01/01/S5P_L2_AER_AI_20230101T000319_20230101T014449_27036.json`, and got a ZeroDivisionError when calling it *after* `fix_item`.


```python
# import sys
# sys.path.append("datasets/sentinel-5p/")

# from sentinel_5p import *
from pctasks.core.storage import StorageFactory
from pctasks.core.tokens import *
from pctasks.core.models.tokens import *
import planetary_computer
import pystac
import shapely.geometry
from stactools.core.utils.antimeridian import fix_item, enclose_poles, Strategy

token = planetary_computer.sas.get_token("sentinel5euwest", "sentinel-5p-stac").token

storage_factory = StorageFactory(tokens=Tokens({"sentinel5euwest": StorageAccountTokens(containers={"sentinel-5p-stac": ContainerTokens(token=token)})}))

asset_uri = "blob://sentinel5euwest/sentinel-5p-stac/TROPOMI/L2__AER_AI/2023/01/01/S5P_L2_AER_AI_20230101T000319_20230101T014449_27036.json"

storage, json_path = storage_factory.get_storage_for_file(asset_uri)
item_dict = storage.read_json(json_path)

item = pystac.Item.from_dict(item_dict)

fix_item(item, Strategy.NORMALIZE)
poly = shapely.geometry.shape(item.geometry)

antimeridian.enclose_poles(poly)  # ZeroDivisionError
```

**Comment by gadomski** [comment]
> Just confirm: we need to add a bit of code here to call the new enclose_poles?

Correct, I'm working this now.

> ZeroDivisionError when calling it after fix_item.

Yeah, enclose poles will have to come first, and I'm doing some testing now to see if we need `fix_item` after it -- I think we will, and that will require some extra twiddling to `enclose_poles` (`fix_item` detects crossings by >180° deltas in longitude, and `enclose_poles` as it is currently written goes straight from meridian to meridian ... we'll need to add some intermediate points to prevent the pole horizontal from getting picked up by `fix_item`).

tl;dr: Antimeridians and poles are hard

File: datasets/sentinel-5p/collection/template.json [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        },
+        {
+            "rel": "about",
+            "href": "https://sentinel.esa.int/web/sentinel/missions/sentinel-5p",
+            "type": "text/html",
+            "title": "Sentinel-5 Precursor Mission"
+        }
+    ],
+    "stac_extensions": [
+        "https://stac-extensions.github.io/sat/v1.0.0/schema.json",
```
Comment (TomAugspurger) [inline datasets/sentinel-5p/collection/template.json:23]: Is this extension necessary? I only see it under `summaries` (I'm not sure if that makes it necessary, but we don't have `eo` for https://planetarycomputer.microsoft.com/api/stac/v1/collections/sentinel-2-l2a).

Comment (TomAugspurger) [reply datasets/sentinel-5p/collection/template.json:23]: The example at https://github.com/radiantearth/stac-spec/blob/fd8fb4e3cf0b6c4a0fe74822accbc1725a92f92a/examples/collection.json does put `eo` in the collection `stac_extension`, despite only using it in summaries. So maybe we need to update our sentinel-2-l2a collection.

Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json:23]: >Is this extension necessary? I only see it under summaries 

My initial thought was, no, it is not necessary since the extension schema does not touch Collection summaries. But the [extension spec](https://github.com/radiantearth/stac-spec/tree/master/extensions) states:

>If a Collection [summary](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#summaries) contains Item fields that implement an extension, then the stac_extensions array of that Collection should list the extension ID.

So, per spec, it should be included. Which aligns with the example Collection you linked to.

**Comment by TomAugspurger** [comment]
> Right now this code is in a personal repo ... if you'd like, I can port it over to stactools before we merge, but I thought I'd get this up now before I do that work. LMK.

I'm OK with running off your personal repo, but agreed that longer-term this belongs in stactools (or shapley).

---

Just for fun, I made a little thing to browse the footprints ([gist](https://gist.github.com/TomAugspurger/bdc7f5694581a112130e5dee9fee1961)). I think these look good.

footprints

**Comment by gadomski** [comment]
> longer-term this belongs in stactools (or shapley).

I'm :-1: on either of these options, unfortunately. **stactools** is too heavy for what the **antimeridian** package does (**antimeridian** only needs **shapely**), and **shapely** intentionally doesn't care about geographic coordinate systems. Maybe there's another good home for this algorithm, but I don't think its either of those spots.

Uhhhh that browser is awesome and I'm going to be using it for sentinel3 _for sure_! Thanks for sharing!

**Comment by gadomski** [comment]
@TomAugspurger released to PyPI and updated PR to point to the released version, so we aren't dependent on my personal repo url anymore.

File: datasets/sentinel-5p/collection/template.json [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            "01",
+            "02",
+            "03"
+        ],
+        "s5p:processing_mode": [
+            "NRTI",
+            "OFFL",
+            "RPRO"
+        ],
+        "s5p:product_type": [
```
Comment (TomAugspurger) [inline datasets/sentinel-5p/collection/template.json:103]: Just to confirm, is this the attribute we expect users to use for filtering products?

It *might* be worth adding a "friendly" product type (maybe call it `s5p:product_name`?) that strips things down a bit (but still keep this one too). I would definitely want to remove the underscroes. Maybe remove `L2`. So

- `AER_AI`
- AER_LH
- CH4
- CLOUD
- CO
- ...

File: datasets/sentinel-5p/collection/template.json [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            "L2__NO2___",
+            "L2__NP_BD3",
+            "L2__NP_BD6",
+            "L2__NP_BD7",
+            "L2__O3_TCL",
+            "L2__O3____",
+            "L2__SO2___"
+        ],
+        "s5p:spatial_resolution": [
+            "5.5x3.5km",
```
Comment (TomAugspurger) [inline datasets/sentinel-5p/collection/template.json:119]: I wonder about the `<horizontal>x<vertical><unit>` pattern here. Is there anything else like this in STAC? Is it better to break that into something more structured?

If we keep this, thoughts on a space vs. no space around the `x` and before the `km`?

**Review by TomAugspurger — APPROVED** [review]
LGTM overall. Just two small questions.

Can you also copy-paste a Dockerfile to the `sentinel-5p` directory (I really need to finish off 
 #104 so we can stop all this duplication)?

File: datasets/sentinel-5p/collection/template.json (hunk 4 above)
Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json:103]: Yes, that was the expectation. But I see your point about that being awkward. I don't have a problem with the stripped down designation. I'll add it.

We should probably do the same for sentinel-3 where we have properties like `"s3:product_type": "SY_2_AOD___"`. In this case we would add an `s3:product_name": "AOD"` property.

File: datasets/sentinel-5p/collection/template.json (hunk 5 above)
Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json:119]: >Is there anything else like this in STAC?

Not directly. There is a `spatial_resolution` field in `raster:bands` (defined as _"Average spatial resolution (in meters) of the pixels in the band."_), but as discussed elsewhere, `raster:bands` is ambiguous as to which band in the NetCDF the information is referencing. 

There is also a `resolution` field in sentinel-3: 
```
      "resolution": [
        1000,
        1000
      ],
```

which is in meters. It's not clear which is which (horizontal vs vertical). But it would be nice to decide on a common solution for sentinel-5 and sentinel-3. I think the list of numbers is nicer than the text string. And specifying distances in meters (not kilometers) is pretty standard in STAC. So maybe a custom spatial resolution field that could go in properties or on an asset?

- `"s5p:spatial_resolution": [5500, 3500]`
- `"s3:spatial_resolution": [1000, 1000]`

Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json]: Done.

File: datasets/sentinel-5p/collection/template.json (hunk 4 above)
Comment (pjhartzell) [reply datasets/sentinel-5p/collection/template.json:103]: Done.

**Comment by TomAugspurger** [comment]
Thanks all!
